### PR TITLE
fix: fix logic to compute service window

### LINKED ISSUE #2017: Inverted service dates in validation report (start later than end)
### Describe the bug

In some validation reports, the service window shows a start date occurring after the end date
Example feeds: [mdb-2138](https://mobilitydatabase.org/feeds/gtfs/mdb-2138) [tld-725](https://mobilitydatabase.org/feeds/gtfs/tld-725)

Here is a screenshot for mdb-2138:

Image

### Steps/Code to Reproduce

See above

### Expected Results

Dates should reflect the contents of dataset.
Note that the dataset could have errors in it, although inversion of data is covered in notice [start_and_end_range_out_of_order-rule](https://gtfs-validator.mobilitydata.org/rules.html#start_and_end_range_out_of_order-rule) and the validation reports with the problem do not have this notice.

### Actual Results

See ab above

### Screenshots

_No response_

### Files used

_No response_

### Validator version

7.00

### Operating system

Irrelevent

### Java version

_No response_

### PR DESCRIPTION
**Pull request by sylvansson**
**Summary:**

- Fix an issue where we shift the service window's start and end dates incorrectly based on exceptions, even if they only apply to a subset of the services (5f001f7fcc4bd381a8f8f667d1c1ef51dd7b5afb).
- Fix an issue where we always calculate the service window based on both calendar.txt and calendar_dates.txt even when one or both are missing (de23cafd5d269f4a060b7f4f3477505c06398a99).
- A few miscellaneous changes (small typos, etc.).

Closes #2017

**Expected behavior:** 

Before:
image

After:
image

I've only tested with the [mdb-2138](https://mobilitydatabase.org/feeds/gtfs/mdb-2138) feed for now. If I understand correctly the acceptance tests don't validate the feed metadata so I still need to check a few datasets manually ⚠️ 

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [ ] Add or update any needed [documentation](https://github.com/MobilityData/gtfs-validator/tree/master/docs) to the repo 
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- [x] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

**Comment by sylvansson** [comment]
I see we're getting a stack overflow for [this dataset](https://storage.googleapis.com/storage/v1/b/mdb-latest/o/de-unknown-basler-verkehrsbetriebe-gtfs-1092.zip). I'll have to take a closer look.

**Comment by sylvansson** [comment]
@davidgamez @jcpitre This is ready for review :-)

File: main/src/main/java/org/mobilitydata/gtfsvalidator/reportsummary/model/ServiceWindow.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import org.mobilitydata.gtfsvalidator.table.GtfsCalendar;
+import org.mobilitydata.gtfsvalidator.table.GtfsCalendarDate;
+import org.mobilitydata.gtfsvalidator.table.GtfsCalendarDateExceptionType;
+import org.mobilitydata.gtfsvalidator.table.GtfsTrip;
+import org.mobilitydata.gtfsvalidator.util.FuncUtil;
+
+record ServiceWindow(LocalDate startDate, LocalDate endDate) {
+  static Optional<ServiceWindow> fromCalendars(
+      List<GtfsTrip> trips, List<GtfsCalendar> allCalendars) {
+    Set<String> serviceIds = new HashSet<>(trips.stream().map(GtfsTrip::serviceId).toList());
```
Comment (davidgamez) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/reportsummary/model/ServiceWindow.java:16]: We had experienced memory performance issues with large feeds, so I recommend replacing this HashSet with an already existing indexing method to perform the lookup by serviceId. As the schema [annotates the serviceId field](https://github.com/MobilityData/gtfs-validator/blob/708856392268751b556fe51d6fa17c9f3f174a3b/main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsTripSchema.java#L36), the `GtfsTripTableContainer` class has a generated method `public List<GtfsTrip> byServiceId(String key)`, which is supported with a map for O(1) items retrieval. We should pass along all ServiceWindow "from" methods the GtfsTripTableContainer and using the indexed method, example:
```
  static Optional<ServiceWindow> fromCalendarDates(
          GtfsTripTableContainer tripTable, List<GtfsCalendarDate> allCalendarDates) {

    List<LocalDate> calendarDates =
        allCalendarDates.stream()
            .filter(
                d ->
                        tripTable.byServiceId(d.serviceId()).size() > 0
                        && d.exceptionType() == GtfsCalendarDateExceptionType.SERVICE_ADDED)
            .map(d -> d.date().getLocalDate())
            .toList();

    // Only empty if there are no calendar dates.
    Optional<LocalDate> startDate = calendarDates.stream().min(LocalDate::compareTo);
    Optional<LocalDate> endDate = calendarDates.stream().max(LocalDate::compareTo);
    return startDate.map(d -> new ServiceWindow(d, endDate.get()));
  }
```

File: main/src/main/java/org/mobilitydata/gtfsvalidator/util/FuncUtil.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package org.mobilitydata.gtfsvalidator.util;
+
+import static java.util.Collections.emptySet;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+public class FuncUtil {
```
Comment (davidgamez) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/util/FuncUtil.java:9]: [suggestion]: Rename it to a more specific scope like, `SetUtils`. This will prevent developers from placing unrelated utility methods in the class.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/util/FuncUtil.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  public static <T> Set<T> intersectAll(Collection<Set<T>> sets) {
```
Comment (davidgamez) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/util/FuncUtil.java:10]: Please add method comments for better documentation.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/reportsummary/model/ServiceWindow.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+import java.time.LocalDate;
+import java.util.*;
+import org.mobilitydata.gtfsvalidator.table.GtfsCalendar;
+import org.mobilitydata.gtfsvalidator.table.GtfsCalendarDate;
+import org.mobilitydata.gtfsvalidator.table.GtfsCalendarDateExceptionType;
+import org.mobilitydata.gtfsvalidator.table.GtfsTrip;
+import org.mobilitydata.gtfsvalidator.util.FuncUtil;
+
+record ServiceWindow(LocalDate startDate, LocalDate endDate) {
+  static Optional<ServiceWindow> fromCalendars(
```
Comment (davidgamez) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/reportsummary/model/ServiceWindow.java:14]: Please add comments to class methods to clarify expected returns like optional empty due to missing or empty files.

**Comment by davidgamez** [comment]
Adding the acceptance test manually due to #1960 (Updated 2026/03/03 )
## 📝 Acceptance Test Report

### 📋 Summary
✅ The rule acceptance has passed for commit a3139d523d642f6e6f75ec811db8d5bfaf76e0be
Download the full acceptance test report [here](https://github.com/MobilityData/gtfs-validator/actions/runs/22642494331) (report will disappear after 90 days).
### 📊 Notices Comparison
<details>
<summary><strong>New Errors</strong> (0 out of 989 datasets, ~0%) ✅</summary>
<p>No changes were detected due to the code change.</p>
</details>
<details>
<summary><strong>Dropped Errors</strong> (0 out of 989 datasets, ~0%) ✅</summary>
<p>No changes were detected due to the code change.</p>
</details>
<details>
<summary><strong>New Warnings</strong> (0 out of 989 datasets, ~0%) ✅</summary>
<p>No changes were detected due to the code change.</p>
</details>
<details>
<summary><strong>Dropped Warnings</strong> (0 out of 989 datasets, ~0%) ✅</summary>
<p>No changes were detected due to the code change.</p>
</details>

### 🛡️ Corruption Check
<details>
<summary><strong>14 out of 1003 sources (~1 %) are corrupted.</summary>

| Dataset   | Ref Report Exists | Ref Report Readable | Latest Report Exists | Latest Report Readable |
|-----------|-------------------|---------------------|----------------------|------------------------|
| mdb-1114 | ✅ | ✅ | ✅ | ✅ |
| mdb-1123 | ✅ | ✅ | ✅ | ✅ |
| mdb-1808 | ✅ | ✅ | ✅ | ✅ |
| mdb-1953 | ✅ | ✅ | ✅ | ✅ |
| mdb-227 | ✅ | ✅ | ✅ | ✅ |
| mdb-383 | ✅ | ✅ | ✅ | ✅ |
| mdb-606 | ✅ | ✅ | ✅ | ✅ |
| mdb-609 | ✅ | ✅ | ✅ | ✅ |
| mdb-655 | ✅ | ✅ | ✅ | ✅ |
| mdb-780 | ✅ | ✅ | ✅ | ✅ |
| mdb-789 | ✅ | ✅ | ✅ | ✅ |
| mdb-806 | ✅ | ✅ | ✅ | ✅ |
| mdb-9 | ✅ | ✅ | ✅ | ✅ |
| mdb-907 | ✅ | ✅ | ✅ | ✅ |
</details>

### ⏱️ Performance Assessment

<details>
<summary><strong>📈 Validation Time</strong></summary>
<p>Assess the performance in terms of seconds taken for the validation process.</p>

| Time Metric                      | Dataset ID        | Reference (s)  | Latest (s)     | Difference (s) |
|-----------------------------|-------------------|----------------|----------------|----------------|
| Average | -- | 5.55 | 5.85 | ⬆️+0.31 |
| Median | -- | 1.62 | 1.92 | ⬆️+0.30 |
| Standard Deviation | -- | 25.29 | 25.34 | ⬆️+0.05 |
| Minimum in References Reports | mdb-2360 | 0.49 | 0.73 | ⬆️+0.24 |
| Maximum in Reference Reports | mdb-2014 | 723.07 | 725.21 | ⬆️+2.14 |
| Minimum in Latest Reports | mdb-1105 | 0.50 | 0.54 | ⬆️+0.04 |
| Maximum in Latest Reports | mdb-2014 | 723.07 | 725.21 | ⬆️+2.14 |
</details>

<details>
<summary><strong>📜 Memory Consumption</strong></summary>

| Metric                      | Dataset ID        | Reference (s)  | Latest (s)     | Difference (s) |
|-----------------------------|-------------------|----------------|----------------|----------------|
| Average | -- | 552.85 MiB | 572.64 MiB | ⬆️+19.79 MiB |
| Median | -- | 325.02 MiB | 325.93 MiB | ⬆️+930.59 KiB |
| Standard Deviation | -- | 970.35 MiB | 1.00 GiB | ⬆️+54.50 MiB |
| Minimum in References Reports | mdb-2297 | 41.30 MiB | 39.55 MiB | ⬇️-1.75 MiB |
| Maximum in Reference Reports | mdb-2014 | 9.10 GiB | 9.11 GiB | ⬆️+8.78 MiB |
| Minimum in Latest Reports | mdb-2297 | 41.30 MiB | 39.55 MiB | ⬇️-1.75 MiB |
| Maximum in Latest Reports | mdb-2014 | 9.10 GiB | 9.11 GiB | ⬆️+8.78 MiB |
</details>

**Review by davidgamez — COMMENTED** [review]
Great work! I added a few comments before final approval.

**Comment by davidgamez** [comment]
@skalexch can you take a look at this PR for product approval? Thanks!

File: main/src/main/java/org/mobilitydata/gtfsvalidator/reportsummary/model/ServiceWindow.java (hunk 1 above)
Comment (sylvansson) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/reportsummary/model/ServiceWindow.java]: That's a great call-out, thank you!

**Comment by davidgamez** [comment]
Adding the acceptance test manually due to https://github.com/MobilityData/gtfs-validator/issues/1960

The test is failing due to a high rate of corrupted files, with no concerns regarding notices. This should be fixed after #2024
## 📝 Acceptance Test Report
### 📋 Summary
❌ The rule acceptance test has failed for commit cdd15beced99c5829e55f8c6bbd39ab17954ed0b
Download the full acceptance test report [here](https://github.com/MobilityData/gtfs-validator/actions/runs/15152307919) (report will disappear after 90 days).
### 📊 Notices Comparison
<details>
<summary><strong>New Errors</strong> (0 out of 1886 datasets, ~0%) ✅</summary>
<p>No changes were detected due to the code change.</p>
</details>
<details>
<summary><strong>Dropped Errors</strong> (0 out of 1886 datasets, ~0%) ✅</summary>
<p>No changes were detected due to the code change.</p>
</details>
<details>
<summary><strong>New Warnings</strong> (0 out of 1886 datasets, ~0%) ✅</summary>
<p>No changes were detected due to the code change.</p>
</details>
<details>
<summary><strong>Dropped Warnings</strong> (0 out of 1886 datasets, ~0%) ✅</summary>
<p>No changes were detected due to the code change.</p>
</details>

### 🛡️ Corruption Check
<details>
<summary><strong>45 out of 1931 sources (~2 %) are corrupted.</summary>

| Dataset   | Ref Report Exists | Ref Report Readable | Latest Report Exists | Latest Report Readable |
|-----------|-------------------|---------------------|----------------------|------------------------|
| br-rio-de-janeiro-angra-dos-reis-viacao-senhor-do-bonfirm-2632 | ✅ | ✅ | ✅ | ✅ |
| ca-london-hamilton-street-railway-gtfs-2618 | ✅ | ✅ | ✅ | ✅ |
| ca-saskatchewan-moose-jaw-transit-gtfs-2602 | ✅ | ✅ | ✅ | ✅ |
| de-baden-wuttemberg-nvbw-gtfs-2393.json | ✅ | ✅ | ✅ | ✅ |
| de-bayern-munchner-verkehrs--und-tarifverbund-gmbh-mvv-gtfs-2252 | ✅ | ✅ | ✅ | ✅ |
| de-sachsen-mitteldeutscher-verkehrsverbund-gmbh-mdv-gtfs-2360 | ✅ | ✅ | ✅ | ✅ |
| es-renfe-alta-velocidad-larga-distancia-media-distancia-gtfs-2620 | ✅ | ✅ | ✅ | ✅ |
| fr-eurostar-gtfs-2431.json | ✅ | ✅ | ✅ | ✅ |
| fr-grand-est-troyes-champagne-metropole-tcat-gtfs-2627 | ✅ | ✅ | ✅ | ✅ |
| fr-grand-est-troyes-champagne-metropole-tcat-gtfs-2628 | ✅ | ✅ | ✅ | ✅ |
| fr-nouvelle-aquitaine-bordeaux-tbm-gtfs-2622 | ✅ | ✅ | ✅ | ✅ |
| fr-sens-intercom-gtfs-2626 | ✅ | ✅ | ✅ | ✅ |
| gg-unknown-sarkshipping-gtfs-2615 | ✅ | ✅ | ✅ | ✅ |
| il-ministry-of-transport-and-road-safety-gtfs-2519 | ✅ | ✅ | ✅ | ✅ |
| jp-aomori-aomori-city-bus-gtfs-2607 | ✅ | ✅ | ✅ | ✅ |
| ro-buzau-transbus-buzau-gtfs-2106 | ✅ | ✅ | ✅ | ✅ |
| ro-dambovita-servicii-publice-municipale-targoviste-gtfs-2107 | ✅ | ✅ | ✅ | ✅ |
| ro-prahova-transport-calatori-express-ploiesti-gtfs-2108 | ✅ | ✅ | ✅ | ✅ |
| us-california-alameda-contra-costa-transit-district-ac-transit-gtfs-2455 | ✅ | ✅ | ✅ | ✅ |
| us-california-county-connection-gtfs-2421 | ✅ | ✅ | ✅ | ✅ |
| us-california-regional-transportation-commission-of-southern-nevada-rtc-gtfs-110 | ✅ | ✅ | ✅ | ✅ |
| us-california-santa-maria-area-transit-gtfs-26 | ✅ | ✅ | ✅ | ✅ |
| us-california-south-county-transit-link-gtfs-2203 | ✅ | ✅ | ✅ | ✅ |
| us-california-taft-maricopa-area-transit-gtfs-821 | ✅ | ✅ | ✅ | ✅ |
| us-california-tri-delta-transit-gtfs-1974 | ✅ | ✅ | ✅ | ✅ |
| us-colorado-vail-blueskylimo-gtfs-2616 | ✅ | ✅ | ✅ | ✅ |
| us-florida-gainesville-regional-transit-system-gtfs-2412 | ✅ | ✅ | ✅ | ✅ |
| us-hawaii-hawaii-mass-transit-agency-hele-on-bus-gtfs-2608 | ✅ | ✅ | ✅ | ✅ |
| us-illinois-danville-mass-transit-gtfs-2363 | ✅ | ✅ | ✅ | ✅ |
| us-kansas-salina-gtfs-1867 | ✅ | ✅ | ✅ | ✅ |
| us-massachusetts-pioneer-valley-transit-authority-pvta-gtfs-2416 | ✅ | ✅ | ✅ | ✅ |
| us-oregon-burns-paiute-tribal-transit-gtfs-2621 | ✅ | ✅ | ✅ | ✅ |
| us-utah-cache-valley-gtfs-1906 | ✅ | ✅ | ✅ | ✅ |
| us-virginia-arlington-transit-gtfs-485 | ✅ | ✅ | ✅ | ✅ |
| us-virginia-fairfax-cue-bus-cue-gtfs-2609 | ✅ | ✅ | ✅ | ✅ |
| us-virginia-fredericksburg-regional-transit-gtfs-2430 | ✅ | ✅ | ✅ | ✅ |
| us-washington-coast-transportation-gtfs-2162 | ✅ | ✅ | ✅ | ✅ |
| us-washington-columbia-county-public-transportation-gtfs-2168 | ✅ | ✅ | ✅ | ✅ |
| us-washington-community-in-motion-gtfs-2163 | ✅ | ✅ | ✅ | ✅ |
| us-washington-eastside-friends-of-seniors-gtfs-2166 | ✅ | ✅ | ✅ | ✅ |
| us-washington-hopelink-transportation-gtfs-2167 | ✅ | ✅ | ✅ | ✅ |
| us-washington-intercity-transit-gtfs-2289 | ✅ | ✅ | ✅ | ✅ |
| us-washington-paratransit-services-gtfs-2176 | ✅ | ✅ | ✅ | ✅ |
| us-washington-puget-sound-educational-service-district-gtfs-2177 | ✅ | ✅ | ✅ | ✅ |
| us-washington-sound-generations-hyde-shuttle-gtfs-2183 | ✅ | ✅ | ✅ | ✅ |
</details>

### ⏱️ Performance Assessment

<details>
<summary><strong>📈 Validation Time</strong></summary>
<p>Assess the performance in terms of seconds taken for the validation process.</p>

| Time Metric                      | Dataset ID        | Reference (s)  | Latest (s)     | Difference (s) |
|-----------------------------|-------------------|----------------|----------------|----------------|
| Average | -- | 3.73 | 3.82 | ⬆️+0.09 |
| Median | -- | 1.38 | 1.50 | ⬆️+0.12 |
| Standard Deviation | -- | 10.68 | 10.33 | ⬇️-0.35 |
| Minimum in References Reports | us-california-san-juan-capistrano-free-weekend-trolley-gtfs-2235 | 0.48 | 0.57 | ⬆️+0.10 |
| Maximum in Reference Reports | gb-unknown-uk-aggregate-feed-gtfs-2014 | 285.52 | 263.50 | ⬇️-22.02 |
| Minimum in Latest Reports | us-california-taft-area-transit-gtfs-2236 | 0.48 | 0.57 | ⬆️+0.09 |
| Maximum in Latest Reports | gb-unknown-uk-aggregate-feed-gtfs-2014 | 285.52 | 263.50 | ⬇️-22.02 |
</details>

<details>
<summary><strong>📜 Memory Consumption</strong></summary>

| Metric                      | Dataset ID        | Reference (s)  | Latest (s)     | Difference (s) |
|-----------------------------|-------------------|----------------|----------------|----------------|
| Average | -- | 459.73 MiB | 460.65 MiB | ⬆️+941.43 KiB |
| Median | -- | 335.93 MiB | 332.04 MiB | ⬇️-3.88 MiB |
| Standard Deviation | -- | 732.24 MiB | 728.43 MiB | ⬇️-3.80 MiB |
| Minimum in References Reports | us-colorado-boulder-county-gtfs-2191 | 38.79 MiB | 411.93 MiB | ⬆️+373.13 MiB |
| Maximum in Reference Reports | ch-unknown-swiss-federal-railways-sbb-gtfs-2144 | 8.37 GiB | 8.03 GiB | ⬇️-345.00 MiB |
| Minimum in Latest Reports | ro-vrancea-consiliul-judetean-vrancea-gtfs-1984 | 39.77 MiB | 37.99 MiB | ⬇️-1.78 MiB |
| Maximum in Latest Reports | ch-unknown-swiss-federal-railways-sbb-gtfs-2144 | 8.37 GiB | 8.03 GiB | ⬇️-345.00 MiB |
</details>

**Comment by sylvansson** [comment]
I see a large increase in memory consumption for the "Minimum in References Reports" metric but other PRs ([this one](https://github.com/MobilityData/gtfs-validator/pull/1988), for example) have the same thing, so I assume it's nothing to worry about?

**Review by skalexch — CHANGES_REQUESTED** [review]
Sorry for not having submitted this earlier, I wrote the comment on Wednesday and did not submit the review. I just wrote one question regarding the logic.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/reportsummary/model/ServiceWindow.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+      GtfsTripTableContainer tripTable,
+      List<GtfsCalendar> calendars,
+      List<GtfsCalendarDate> calendarDates) {
+    Optional<ServiceWindow> serviceWindowFromCalendars =
+        ServiceWindow.fromCalendars(tripTable, calendars);
+    if (serviceWindowFromCalendars.isEmpty()) {
+      return Optional.empty();
+    }
+
+    Map<String, Set<LocalDate>> removedDaysByServiceId =
```
[the remark this answers is not in the document]
Comment (davidgamez) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/reportsummary/model/ServiceWindow.java:78]: Thanks, @skalexch. You are right. The spec doesn't force all service IDs to be defined in the calendar.txt and calendar_dates.txt when both files are defined. In this case, we can have service IDs from calendar.txt that are not present in the calendar_dates and will have a service on a day removed by a calendar_date entry. 

Following this logic, I agree that we need to keep a total of "all unique service IDs" per day, so we make sure that all services are not present to mark the date as "removed."  The "all unique service IDs" can be computed by looking at the trips.txt file.
In addition, we need to ensure that the added service "expands" the start and end dates of the service window. So, we must iterate over all additions and set the start and end dates accordingly. This should be done after computing the start and end dates from all calendar ranges.

**Review by davidgamez — CHANGES_REQUESTED** [review]

File: main/src/main/java/org/mobilitydata/gtfsvalidator/reportsummary/model/ServiceWindow.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sylvansson) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/reportsummary/model/ServiceWindow.java:78]: Good catch, I'll try to get this fixed sometime next week.

**Comment by emmambd** [comment]
Looks like this feed isn't wrong, our service date window is. Should test the fix with it: https://mobilitydatabase.org/feeds/gtfs/tdg-81549

**Comment by skalexch** [comment]
@emmambd I found this service period in the feed: 3983,1,1,1,1,1,0,1,20231216,52020430

**Comment by sylvansson** [comment]
Thanks @emmambd. I'll try to address David's comments within the next two weeks, sorry for the delay!

**Review by sylvansson — COMMENTED** [review]
Apologies for the very, _very_ long delay.

**Comment by sylvansson** [comment]
We ignore added dates if they fall outside of all the "regular" service ranges. For example, if the latest end date in calendar.txt is 2026-01-01 and there's an exception in calendar_dates.txt that adds service on 2026-02-30 (two months later), we keep 2026-01-01 as the end date. I assume this is what we want because calendar dates should be treated as exceptions when both files are present, but I just want to confirm.

> The [calendar_dates.txt](https://gtfs.org/documentation/schedule/reference/#calendar_datestxt) table explicitly activates or disables service by date. It may be used in two ways.
>
> Recommended: Use [calendar_dates.txt](https://gtfs.org/documentation/schedule/reference/#calendar_datestxt) in conjunction with [calendar.txt](https://gtfs.org/documentation/schedule/reference/#calendartxt) to define exceptions to the default service patterns defined in [calendar.txt](https://gtfs.org/documentation/schedule/reference/#calendartxt). If service is generally regular, with a few changes on explicit dates (for instance, to accommodate special event services, or a school schedule), this is a good approach. In this case `calendar_dates.service_id` is a foreign ID referencing `calendar.service_id`.

**Comment by skalexch** [comment]
@sylvansson I think added dates should be considered. For some feeds, the service window is (unfortunately) fully defined by added dates. Theoretically, the service window should span the period from the service first operates until it last operates in the feed. Happy to hear any pushback on that though!

**Comment by davidgamez** [comment]
Hi @sylvansson, to make sure we’re covering all the edge cases, I opened sylvansson/gtfs-validator#1 against this branch with a more extensive set of unit tests for ServiceWindow.
At this point, ServiceWindow really deserves its own thorough test class 😄. Most scenarios pass, but two tests are currently failing because the current implementation doesn’t support those cases:
- get_firstWednesdayRemoved_startStillOnFirstThursday: expects the window start to move to the first Thursday with service when the first Wednesday in the range is fully removed. The implementation currently ignores weekday patterns and only uses the calendar date range, so it doesn’t shift the start as expected.
- get_bothTables_addedDateOutsideCalendarRange_definesWindow: expects a SERVICE_ADDED date that lies completely outside the calendar.txt range (for a service that has trips) to extend the service window. The current fromCalendarsAndCalendarDates logic only derives min/max from calendar-based dates and doesn’t incorporate added-only dates outside that range.
I think we should support both behaviors, but I left the implementation unchanged for now so the tests highlight the gap.

Also, I was considering making fromCalendarsAndCalendarDates(...) and fromCalendarDates(...) private helpers. Right now get(...) is the only intended public entry point (it applies the trip-based filtering and chooses the correct path), and hiding the helpers would avoid external callers depending on their current internal behavior while we iterate on edge cases like the failing tests.

**Comment by sylvansson** [comment]
@davidgamez I made a few small changes to the tests you added in https://github.com/sylvansson/gtfs-validator/pull/1; see 86568cc2ee984d6b5806f50330337bfad7eba093.

**Comment by sylvansson** [comment]
We now take weekday patterns into account, as well as ADDED exceptions even if they're outside the service interval. I also made the helper functions private.
